cpu/stm32: fix riotboot settings for L4 and WB

gschorcht opened the pull request:
### Contribution description

This PR fixes the `riotboot` configuration for L4 and WB.

The family is not called `stm32l4` or `stm32wb` but `l4` and `wb`. That is, the `riotboot` configuration didn't work at all. Furthermore, a minimum `RIOTBOOT_LEN` of `0x2000` is required for L4.

Found when investigating the compilation errors for `bootloaders/riotboot_serial` in PR #19576.

### Testing procedure

1. Green CI.
2. Use the following commands:
    ```
    BOARD=nucleo-l496zg make -C tests/riotboot info-debug-variable-RIOTBOOT_HDR_LEN
    BOARD=p-nucleo-wb55 make -C tests/riotboot info-debug-variable-RIOTBOOT_HDR_LEN
    ```
    In master these commands give
    ```
    0x400
    ```
    With this PR these commands give
    ```
    0x200
    ```
    as expected.
3. Use the following commands:
    ```
    BOARD=nucleo-l496zg make -C tests/riotboot info-debug-variable-RIOTBOOT_LEN
    BOARD=p-nucleo-wb55 make -C tests/riotboot info-debug-variable-RIOTBOOT_LEN
    ```
    In master these commands give
    ```
    0x1000
    ```
    With this PR these commands give
    ```
    0x2000
    ```
    as expected.

### Issues/PRs references

aabadie commented:
I tested `tests/riotboot` with this PR on nucleo-l476rg and p-l496g-cell02: it works for both. But it also works on master so maybe I'm not testing the right thing ?

aabadie commented on cpu/stm32/stm32_riotboot.mk:
```diff
@@ -25,16 +25,21 @@ ifneq (,$(filter $(CPU_FAM),f2 f4 f7))
+else ifeq (l4,$(CPU_FAM))
   # "The Vector table must be naturally aligned to a power of two whose alignment
   # value is greater than or equal to number of Exceptions supported x 4"
-  # CPU_IRQ_NUMOFF for stm32l4 boards is < 91+16 so (107*4 bytes = 428 bytes ~= 0x200)
+  # CPU_IRQ_NUMOFF for stm32l4 boards is < 95+16 so (111*4 bytes = 444 bytes ~= 0x200)
   # RIOTBOOT_HDR_LEN can be set to 0x200
   RIOTBOOT_HDR_LEN ?= 0x200
-else ifeq (stm32wb,$(CPU_FAM))
+
+  # Slot size is determined by "((total_flash_size - RIOTBOOT_LEN) / 2)".
+  # If RIOTBOOT_LEN uses an odd number of flashpages, the remainder of the
+  # flash cannot be divided by two slots while staying FLASHPAGE_SIZE aligned.
+  RIOTBOOT_LEN ?= 0x2000
+else ifeq (wb,$(CPU_FAM))
   # "The Vector table must be naturally aligned to a power of two whose alignment
   # value is greater than or equal to number of Exceptions supported x 4"
-  # CPU_IRQ_NUMOFF for stm32l4 boards is < 91+16 so (107*4 bytes = 428 bytes ~= 0x200)
+  # CPU_IRQ_NUMOFF for stm32wb boards is < 63+16 so (79*4 bytes = 316 bytes ~= 0x200)
   # RIOTBOOT_HDR_LEN can be set to 0x200
   RIOTBOOT_HDR_LEN ?= 0x200
```
Could the 2 cases be merged in a single one, like this:

```mk
else ifneq (,$(filter l4 wb,$(CPU_FAM)))
  # "The Vector table must be naturally aligned to a power of two whose alignment
  # value is greater than or equal to number of Exceptions supported x 4"
  # CPU_IRQ_NUMOFF for stm32l4 boards is < 95+16 so (111*4 bytes = 444 bytes ~= 0x200)
  # CPU_IRQ_NUMOFF for stm32wb boards is < 63+16 so (79*4 bytes = 316 bytes ~= 0x200)
  # RIOTBOOT_HDR_LEN can be set to 0x200
  RIOTBOOT_HDR_LEN ?= 0x200

  # Slot size is determined by "((total_flash_size - RIOTBOOT_LEN) / 2)".
  # If RIOTBOOT_LEN uses an odd number of flashpages, the remainder of the
  # flash cannot be divided by two slots while staying FLASHPAGE_SIZE aligned.
  RIOTBOOT_LEN ?= 0x2000
endif
```

gschorcht commented:
> I tested `tests/riotboot` with this PR on nucleo-l476rg and p-l496g-cell02: it works for both. But it also works on master so maybe I'm not testing the right thing ?

It works since it used the default value of `0x400` for `RIOTBOOT_HDR_LEN` set in `cpu/cortex_common/Makefile.include`. You can use
```
BOARD=nucleo-l496zg make -C tests/riotboot info-debug-variable-RIOTBOOT_HDR_LEN
```
to verify.

gschorcht commented:
I have changed as suggested and squashed it directly.

To check you should use the following commands:
```
BOARD=nucleo-l496zg make -C tests/riotboot info-debug-variable-RIOTBOOT_HDR_LEN
BOARD=p-nucleo-wb55 make -C tests/riotboot info-debug-variable-RIOTBOOT_HDR_LEN
BOARD=nucleo-l496zg make -C tests/riotboot info-debug-variable-RIOTBOOT_LEN
BOARD=p-nucleo-wb55 make -C tests/riotboot info-debug-variable-RIOTBOOT_LEN
```
You should observe the wrong values are given in master.

aabadie reviewed (approved):
On L4 the number of IRQs varies from 83 (L412) to 95 (L4R5 and the like). So 0x200 is enough in all cases.

I tested the values are correct following https://github.com/RIOT-OS/RIOT/pull/19618#issuecomment-1557420555

ACK

aabadie commented:
bors merge

aabadie commented:
bors cancel

aabadie commented:
bors merge

aabadie commented:
Hmm, Murdock spotted a problem with `nucleo-l4r5zi`

gschorcht commented:
> Hmm, Murdock spotted a problem with `nucleo-l4r5zi`

Due to the bug, `RIOTBOOT_LEN` wasn't set to `0x2000` here https://github.com/RIOT-OS/RIOT/blob/8408d69ea8baf93e91a4312891e09f3b249b572c/cpu/stm32/stm32_riotboot.mk#L44 but in `cpu/cortexm_common`: https://github.com/RIOT-OS/RIOT/blob/8408d69ea8baf93e91a4312891e09f3b249b572c/cpu/cortexm_common/Makefile.include#L73-L77 Therefore, `RIOTBOOT_LEN` was set to `0x4000` in case `usbus_dfu` or `tinyusb_dfu` are used. It seems that we also need deal with this case for L4 and WB.

aabadie commented:
bors cancel

gschorcht commented:
> Therefore, `RIOTBOOT_LEN` was set to `0x4000` in case `usbus_dfu` or `tinyusb_dfu` are used. It seems that we also need to deal with this case for L4 and WB.

However, I wonder why we should increase `RIOTBOOT_LEN` to 8 KByte for all WB and L4 boards, when it worked with the bug even with only 4 kByte, especially for MCUs with only 128 kByte of flash. The problem appeared only with `stm32l496g-disco`, because it must use additionally the module `periph_lpuart` and 4 KByte as `RIOTBOOT_LEN` are just 200 byte too little. 

On the other hand, I'm not sure whether it would make sense to define `RIOTBOOT_LEN` on board level.

gschorcht commented:
I have changed it back for L4 so that `RIOTBOOT_LEN` is set to the default values as defined in `cpu/cortexm_common/Makefile.include`. I fixed `RIOTBOOT_LEN` for WB. The problem with `stm32l496g-disco` is solved on board level.

benpicco commented:
bors merge

gschorcht commented:
Thanks for reviewing and merging :smile:
